posthog/mixins.py

```diff
@@ -1,6 +1,111 @@
+    left = iter(t)
+    right = iter(t[1:])
+    return zip_longest(left, right)
+
+
+def _version_from_request(instance, state: str, request, action: str) -> None:
+    instance_state = json.loads(state)
+    id = instance_state["pk"]  # because deleted instances have no primary key
+
+    version = HistoricalVersion(
+        state=state,
+        name=instance.__class__.__name__,
+        action=action,
+        item_id=id,
+        created_by_name=request.user.first_name,
+        created_by_email=request.user.email,
+        created_by_id=request.user.id,
+    )
+
+    if hasattr(instance, "team_id"):
+        version.team_id = instance.team_id
+
+    version.save()
+
+
+def _version_from_serializer(serializer, action: str) -> None:
+    HistoricalVersion(
+        state=serializer.data,
+        name=serializer.instance.__class__.__name__,
+        item_id=serializer.instance.id,
+        action=action,
+        created_by_name=serializer.context["request"].user.first_name,
+        created_by_email=serializer.context["request"].user.email,
+        created_by_id=serializer.context["request"].user.id,
+        team_id=serializer.context["team_id"],
+    ).save()
+
+
+class HistoryListItemSerializer(serializers.Serializer):
+    email = serializers.EmailField(read_only=True)
+    name = serializers.CharField(read_only=True)
+    user_id = serializers.IntegerField(read_only=True)
+    action = serializers.CharField(read_only=True)
+    detail = serializers.DictField(read_only=True)
+    created_at = serializers.CharField(read_only=True)
+
+
+def _should_log_history(instance: Any) -> bool:
+    return instance.__class__.__name__ in ["FeatureFlag"]
+
+
+class HistoryLoggingMixin:
+    def perform_create(self, serializer):
+        serializer.save()
+        _version_from_serializer(serializer, "create")
+
+    def perform_update(self, serializer):
+        serializer.save()
+        _version_from_serializer(serializer, "update")
+
+    @action(methods=["GET"], detail=True)
+    def history(self, request: request.Request, **kwargs):
+        """
+        Because we're inside a mixin being applied to a viewset
+        we can add a history endpoint here to avoid cluttering up the viewset with history code
+
+        so `self` here is the viewset
+
+        this
+         * uses the viewset's object to determine what type it is operating on,
+         * loads a page of history for that type and id (newest first)
+         * and returns a computed history for that page of history
+        """
+        history_type = self.get_object().__class__.__name__  # type: ignore
+        # in order to make a page of up to 10 we need to get up to 11 as we need N-1 to determine what changed
+        versions = HistoricalVersion.objects.filter(
+            team_id=self.team.id, name=history_type, item_id=kwargs["pk"]  # type: ignore
+        ).order_by("-versioned_at")[:11]
```

Comment:
This `10+1` could be a query param perhaps, but all in due time...

Reply:
Yep, I'm assuming paging follows up soon after